### PR TITLE
Prometheus: Add $__rate_interval_ms to go along with $__interval_ms

### LINKED ISSUE #69581: Feature: add $__rate_interval_ms variable for numerical division
### 
# What went wrong?

There are a few useful built-in grafana-specific variables, like __interval and __rate_interval.  For some calculations, it's necessary to do division on an interval value, and for __interval there is __interval_ms: https://github.com/grafana/grafana/blob/main/pkg/tsdb/prometheus/models/query.go#L27-L34

However, there is no __rate_interval_ms.  For some of our internal team work, we'd like to add __rate_interval_ms as a similar numeric value 

### 
# How do we reproduce it?

$__rate_interval_ms simply does not exist

### 
# What Grafana version are you using?

dev

### 
### Is the bug inside a Dashboard Panel?

_No response_

### Grafana Platform?

I use Grafana Cloud

### User's OS?

_No response_

### User's Browser?

_No response_

### Is this a Regression?

None

### Are Datasources involved?

Prometheus

### Anything else to add?

_No response_

### PR DESCRIPTION
**Pull request by ywwg**
**What is this feature?**

Adds $__rate_interval_ms variable to go along with existing __interval_ms variable

**Why do we need this feature?**

It is sometimes necessary to calculate rates using numeric division rather than a duration, so the _ms version is very useful.

**Who is this feature for?**

Query team automated translations

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/69581

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

Not sure which docs might need updating

### CODE REVIEW

**Comment by gtk-grafana** [comment]
A few questions on this:
`rate(tempo_query_frontend_queries_total{cluster="dev-us-central-0"}[$__interval_msms])` is the only way I've found of using this, do you have some examples so we can better understand the use case for this? Taking a look at the examples in the unit tests isn't helpful as those don't appear to be valid promQL expressions.

Also these variables don't show up in the frontend 
image
image
Do we want them to?

**Comment by gtk-grafana** [comment]
Also we need to add this to the docs: https://grafana.com/docs/grafana/latest/dashboards/variables/add-template-variables/

**Comment by ywwg** [comment]
$__interval_msms is incorrect, it should be $__interval_ms.  Not sure why ms is doubling?

The use-case for this is datadog query translations where a ddog query like: `per_hour(sum:foo{bar:baz} by {blip})` will become `3600 * sum by (blip) (clamp_min(deriv(((sum_over_time(foo{_dot_internal_dot_dd__type="rate",bar=~".*'baz'.*"}[$__rate_interval]) / ($__rate_interval_ms * 1000)))[$__rate_interval:$__rate_interval]), 0))`

We can use $__interval but that's not always the right value

Yes the variable should show up in the UI but I don't know where to do that

File: docs/sources/dashboards/variables/add-template-variables/index.md [hunk 1]
```diff
@@ -314,6 +314,10 @@ Currently only supported for Prometheus and Loki data sources. This variable rep
 
 Currently only supported for Prometheus data sources. The `$__rate_interval` variable is meant to be used in the rate function. Refer to [Prometheus query variables]({{< relref "../../../datasources/prometheus/template-variables#use-__rate_interval" >}}) for details.
 
+### $\_\_rate_interval_ms
+
+This variable is the `$__rate_interval` variable in milliseconds, not a time interval formatted string. For example, if the `$__rate_interval` is `20m` then the `$__rate_interval_ms` is `1200000`.
```
Comment (imatwawana) [inline docs/sources/dashboards/variables/add-template-variables/index.md:319]:
```suggestion
This variable is the `$__rate_interval` variable in milliseconds, not a time-interval-formatted string. For example, if the `$__rate_interval` is `20m` then the `$__rate_interval_ms` is `1200000`.
```

**Review by imatwawana — APPROVED** [review]
One nit, but otherwise LGTM!

File: docs/sources/dashboards/variables/add-template-variables/index.md (hunk 1 above)
Comment (gtk-grafana) [reply docs/sources/dashboards/variables/add-template-variables/index.md:319]: Is this supported by prometheus and Loki? Can we add that to the doc?

**Comment by gtk-grafana** [comment]
Ok sorry about that, here's the proof of concept: https://github.com/grafana/grafana/pull/72303. 

I don't think we should pull this into the autocomplete in the frontend, and your PR is fine to go as is.

The reasoning being that this new variable will be displayed in any context the query editor is expecting a rate range.
image
Which suffers from the problem that in this context we're expecting a time unit along with the number
image
And then you wind up doing this:
image
But in this context it will give the same output as just using $__rate_interval.

In the use case you provided:
Using promQL expression of `3600 * sum by (blip) (clamp_min(deriv(((sum_over_time(go_gc_duration_seconds_count[$__rate_interval]) / ($__rate_interval_ms * 1000)))[$__rate_interval:$__rate_interval]), 0))` query with `gdev-prometheus` devenv.

image

The current promql parser has no idea what you're doing in there, so this is unlikely to benefit folks in autocomplete as much as it is going to confuse users. 

If you want to create an issue and assign to @grafana/observability-metrics so we have this decision documented that would be awesome.

File: docs/sources/dashboards/variables/add-template-variables/index.md (hunk 1 above)
Comment (ywwg) [reply docs/sources/dashboards/variables/add-template-variables/index.md:319]: I believe variables are all handled in Grafana and don't require specific support in the back ends. So, this should be handled by all of them.

**Comment by gtk-grafana** [comment]
Yeah, this is a new feature so we don't need to backport it, I think the only thing left to resolve is it's saying there's a [formatting error](https://drone.grafana.net/grafana/grafana/128447/5/5) in the doc and we should be good to go

```
+ yarn run prettier:checkDocs
[warn] docs/sources/dashboards/variables/add-template-variables/index.md
[warn] Code style issues found in the above file. Run Prettier to fix.
```

File: docs/sources/dashboards/variables/add-template-variables/index.md [hunk 2]
```diff
@@ -1,4 +1,4 @@
----
+****---
```
Comment (itsmylife) [inline docs/sources/dashboards/variables/add-template-variables/index.md:1]:
```suggestion
---
```

**Review by itsmylife — APPROVED** [review]

**Comment by itsmylife** [comment]
I have no idea why it is failing.

**Pull request merged**